### ISSUE TITLE
Research: CosyVoice browser-side TTS prototype results

### ISSUE THREAD

**Issue by jiangzhuo** [kizuna-ai-lab/sokuji#159]
## Summary

Built a prototype to validate CosyVoice 3 (Fun-CosyVoice3-0.5B) running in-browser via ONNX Runtime Web. The full 14-model pipeline (3.8GB) runs end-to-end and produces partially intelligible speech, but is not yet viable for production.

**Branch:** `feat/cosyvoice-webgpu-proto`
**Spec:** `docs/superpowers/specs/2026-03-30-cosyvoice-webgpu-prototype-design.md`

## What CosyVoice offers

- Zero-shot voice cloning with a brief reference audio sample
- 9 languages (zh/en/ja/ko/de/es/fr/it/ru) + 18 Chinese dialects
- 0.5B parameter LLM-based TTS with 150ms streaming latency (on GPU)
- 24kHz output — matches Sokuji's system standard

## Prototype Architecture

```
Ctrl+Shift+V → CosyVoiceProto overlay → Web Worker → 14 ONNX models

Stage 0: Prompt Processing (campplus + speech_tokenizer)
Stage 1: LLM Autoregressive (Qwen2-based, top-k sampling)
Stage 2: Flow Diffusion (DiT, 10 Euler steps, CFG batch=2)
Stage 3: HiFT Vocoder (F0 → source → STFT → decoder → ISTFT)
```

Pure JS audio preprocessing (FFT, 3 mel configs, STFT/ISTFT). Tokenizer via `@huggingface/transformers`. Models from `ayousanz/cosy-voice3-onnx`.

## Test Results

| Stage | Time (WASM) | Notes |
|-------|-------------|-------|
| Prompt Processing | 19.5s | Speaker embedding + speech tokens extracted correctly |
| LLM Autoregressive | 231.9s | 500 tokens @ 2.2 tok/s, never hit EOS |
| Flow Diffusion | 1305.0s | 664MB DiT × 10 steps, main bottleneck |
| Vocoder | 37.9s | Correct pipeline |
| **Total** | **~26 min** | Not viable on WASM |

**Quality:** First several words clearly intelligible. Degrades after ~10-20 tokens due to EOS not being generated (LLM runs to max 500 tokens).

### What works
- Full 4-stage pipeline logic is correct (verified via diagnostic logs)
- Staged WASM loading avoids OOM (create/release sessions per stage)
- Pure JS mel/FFT/STFT matches expected model inputs
- Dynamic input name mapping handles ONNX model variations

### What blocks production use

1. **WebGPU FP16 incompatibility** — All FP16 models produce `Invalid ComputePipeline` errors for Softmax, MatMul, Transpose on GPUs without `shader-f16`. ORT WebGPU EP doesn't gracefully fall back to FP32 compute. This forces all models to WASM (CPU), making it ~100x slower than needed.

2. **EOS never generated** — LLM produces 500 tokens without EOS (token 6562). Possibly FP16→WASM precision loss affecting logit distribution, or temperature/repetition penalty tuning needed.

3. **FP16 graph optimization crash on WASM** — `graphOptimizationLevel: 'all'` causes `InsertedPrecisionFreeCast` errors. Workaround: use `'basic'` for FP16 models.

### Option A: FP32 ONNX model exports
Re-export all 14 models in FP32. WebGPU EP should work correctly with FP32. ~2x model size (7-8GB) but one-time download with IndexedDB caching. Estimated LLM speed: 10-30 tok/s on WebGPU.

### Option B: Server-side deployment
Deploy CosyVoice FastAPI server behind sokuji-backend. Stream audio via existing `onConversationUpdated` pipeline. Simplest integration, guaranteed quality, but requires GPU server and is not offline-capable.

### Option C: Wait for ORT WebGPU FP16 improvements
ONNX Runtime Web's WebGPU backend is actively being improved. Future versions may handle FP16 models without `shader-f16` by auto-casting to FP32. Monitor onnxruntime releases.

### Option D: INT4 quantization
Quantize LLM backbone and flow estimator to INT4 (~1GB total). Smaller download, likely WebGPU compatible. Need to validate quality.

## Files Changed

```
src/lib/local-inference/cosyvoice/
├── types.ts                        # Worker message protocol
├── audio-utils.ts                  # Pure JS FFT, mel, STFT/ISTFT
├── cosyvoice-webgpu.worker.ts      # 14-model inference pipeline
├── CosyVoiceProto.tsx              # Prototype UI overlay
└── CosyVoiceProto.scss             # Dark theme styles

src/App.tsx                         # Ctrl+Shift+V keyboard shortcut
```

**Comment by matarwapporter** [comment]
@jiangzhuo

各位开源社区的伙伴们：

这个仓库的情况（过去 48 小时内新增了约 3192 颗星，没有头像、没有仓库、没有 follower，注册日期集中在最近两周）相信大家都看到了。我想借这个机会跟大家聊几句，不只是关于这一个项目，而是关于我们共同面对的一个挑战。

**致维护者：**
我理解你的压力和焦虑。但刷星不是出路。这条路的终点不是成功，是信任的崩塌。现在回头还来得及——清理数据、公开说明、重新出发。社区会给知错能改的人第二次机会。

**致其他维护者：**
看到这个案例，请引以为戒。不管压力多大，不管别人怎么做，守住底线。真实的十颗 Star 胜过虚假的一万颗。

**致所有用户：**
养成查看 Stargazer 质量的习惯。不要只看数字做决定。如果发现异常，勇敢指出来。你的一次发声可能帮助很多人避坑。

**致平台方：**
刷星问题不是靠社区自治能解决的，需要平台层面的技术支持和制度保障。希望 GitHub 能在这方面投入更多资源。

开源社区是我们共同的家。让我们一起守护它。

社区在看着，不要让大家失望。
